pkg/stanza/operator/transformer/recombine/recombine.go

```diff
@@ -267,73 +277,84 @@ func (r *Transformer) matchIndicatesLast() bool {
 
 // addToBatch adds the current entry to the current batch of entries that will be combined
 func (r *Transformer) addToBatch(_ context.Context, e *entry.Entry, source string) {
-	if _, ok := r.batchMap[source]; !ok {
-		r.batchMap[source] = []*entry.Entry{e}
+	if batch, ok := r.batchMap[source]; !ok {
+		r.batchMap[source] = &sourceBatch{
+			entries:    []*entry.Entry{e},
+			recombined: strings.Builder{},
+		}
 		if len(r.batchMap) >= r.maxSources {
 			r.Error("Batched source exceeds max source size. Flushing all batched logs. Consider increasing max_sources parameter")
 			r.flushUncombined(context.Background())
+			return
 		}
+	} else {
+		batch.entries = append(batch.entries, e)
+	}
+	batch := r.batchMap[source]
```

Comment:
```suggestion
	batch, ok := r.batchMap[source]
	if !ok {
		if len(r.batchMap) == r.maxSources {
			r.Error("Batched source exceeds max source size. Flushing all batched logs. Consider increasing max_sources parameter")
			r.flushUncombined(context.Background())
		}
		batch = &sourceBatch{recombined: strings.Builder{}}
		r.batchMap[source] = batch		
	}
	batch.entries = append(batch.entries, e)
```

Reply:
Thanks for your reminder, but I think the suggested change is not correct because the maxSources limit may be exceeded only when a new batch is created, so if !ok, we should create a batch firstly and then compare len(r.batchMap) and r.maxSources.

Reply:
Currently, it will detect that we've exceeded batch limit immediately after adding a batch. The change I suggested would wait until the next cycle but still detect the same condition before adding the next batch. It's a minor difference but I think you are right that it may cause a problem because we do not know when the next cycle will happen. 

I'm open to other suggestions but would prefer if we find a way to clean up this section somehow. In fact, I think it has a bug currently because `flushUncombined` will reset the entire `batchMap`, which means that `batch := r.batchMap[source]` will then set `batch` to `nil`.

Reply:
I've changed the code. And I'm not very clear about the bug because the function will **return directly** after **flushCombined**.

Reply:
Ah, you're right again. Thanks for double checking me on that too. Anyways, I think what you changed is sufficient because we no longer reinstantiate `batch` unnecessarily.